RESTEasy reactive endpoint 404

### Describe the bug

When two endpoint with similar URI's are created, one of them returns 404. Ex:
/device-features/{featureId}/permissions
/device-features/{featureId}/actions

### Expected behavior

Both Endpoints return 200

### Actual behavior

_No response_

### How to Reproduce?

https://github.com/victor-costa-avancee/resteasy_path_reproducer

### Output of `uname -a` or `ver`

Darwin MBP-de-Vic.localdomain 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.1.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

I have also noticed that when sending an empty body to a POST Method a JsonbException is thrown.
```2021-08-09 12:36:24,412 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Request failed: javax.json.bind.JsonbException: Internal error: Invalid token=EOF at (line no=1, column no=0, offset=-1). Expected tokens are: [CURLYOPEN, SQUAREOPEN, STRING, NUMBER, TRUE, FALSE, NULL]
        at org.eclipse.yasson.internal.Unmarshaller.deserializeItem(Unmarshaller.java:68)
        at org.eclipse.yasson.internal.Unmarshaller.deserialize(Unmarshaller.java:51)
        at org.eclipse.yasson.internal.JsonBinding.deserialize(JsonBinding.java:59)
        at org.eclipse.yasson.internal.JsonBinding.fromJson(JsonBinding.java:99)
        at io.quarkus.resteasy.reactive.jsonb.runtime.serialisers.JsonbMessageBodyReader.doReadFrom(JsonbMessageBodyReader.java:55)
        at io.quarkus.resteasy.reactive.jsonb.runtime.serialisers.JsonbMessageBodyReader.readFrom(JsonbMessageBodyReader.java:38)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.readFrom(RequestDeserializeHandler.java:108)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:68)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:25)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:133)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:536)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.json.stream.JsonParsingException: Invalid token=EOF at (line no=1, column no=0, offset=-1). Expected tokens are: [CURLYOPEN, SQUAREOPEN, STRING, NUMBER, TRUE, FALSE, NULL]
        at org.glassfish.json.JsonParserImpl.parsingException(JsonParserImpl.java:426)
        at org.glassfish.json.JsonParserImpl.access$1100(JsonParserImpl.java:55)
        at org.glassfish.json.JsonParserImpl$NoneContext.getNextEvent(JsonParserImpl.java:414)
        at org.glassfish.json.JsonParserImpl.next(JsonParserImpl.java:352)
        at org.eclipse.yasson.internal.JsonbRiParser.next(JsonbRiParser.java:141)
        at org.eclipse.yasson.internal.Unmarshaller.getRootEvent(Unmarshaller.java:80)
        at org.eclipse.yasson.internal.Unmarshaller.deserializeItem(Unmarshaller.java:58)
        ... 16 more
```

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/PathMatcher.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/PathMatcher.java
@@ -65,25 +65,9 @@ public class PathMatcher<T> implements Dumpable {
             } else if (pathLength < length) {
                 char c = path.charAt(pathLength);
                 if (c == '/') {
-                    //todo: matches that end with / are a bit weird, maybe this is not the best matching algo
-                    SubstringMap.SubstringMatch<T> next = paths.get(path, pathLength + 1);
-                    if (next != null && next.getKey().endsWith("/")) {
-                        return new PathMatch<>(next.getKey(), path.substring(pathLength), next.getValue());
-                    }
-                    next = paths.get(path, pathLength);
+                    SubstringMap.SubstringMatch<T> next = paths.get(path, pathLength);
                     if (next != null) {
                         return new PathMatch<>(next.getKey(), path.substring(pathLength), next.getValue());
-                    }
-                }
-                // it's also possible that we're looking up /foo/bar/gee and have a current path of /foo/bar/ which
-                // is an acceptable prefix
-                if (pathLength > 0) {
-                    c = path.charAt(pathLength - 1);
-                    if (c == '/') {
-                        SubstringMap.SubstringMatch<T> next = paths.get(path, pathLength);
-                        if (next != null) {
-                            return new PathMatch<>(next.getKey(), path.substring(pathLength), next.getValue());
-                        }
                     }
                 }
             }
@@ -111,6 +95,8 @@ public class PathMatcher<T> implements Dumpable {
         if (PathMatcher.STRING_PATH_SEPARATOR.equals(path)) {
             this.defaultHandler = handler;
             return this;
+        } else if (path.endsWith(STRING_PATH_SEPARATOR)) {
+            throw new RuntimeException("Prefix path cannot end with /");
         }
 
         paths.put(path, handler);

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/URITemplate.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/URITemplate.java
@@ -51,10 +51,7 @@ public class URITemplate implements Dumpable, Comparable<URITemplate> {
                         state = 1;
                         if (sb.length() > 0) {
                             String literal = sb.toString();
-                            if (components.isEmpty()) {
-                                stem = literal;
-                            }
-                            components.add(new TemplateComponent(Type.LITERAL, literal, null, null, null));
+                            stem = handlePossibleStem((List<TemplateComponent>) components, stem, literal);
                         }
                         sb.setLength(0);
                     } else {
@@ -113,10 +110,7 @@ public class URITemplate implements Dumpable, Comparable<URITemplate> {
             case 0:
                 if (sb.length() > 0) {
                     String literal = sb.toString();
-                    if (components.isEmpty()) {
-                        stem = literal;
-                    }
-                    components.add(new TemplateComponent(Type.LITERAL, literal, null, null, null));
+                    stem = handlePossibleStem(components, stem, literal);
                 }
                 break;
             case 1:
@@ -169,6 +163,26 @@ public class URITemplate implements Dumpable, Comparable<URITemplate> {
         this.capturingGroups = capGroups;
         this.complexExpressions = complexGroups;
 
+    }
+
+    private String handlePossibleStem(List<TemplateComponent> components, String stem, String literal) {
+        if (components.isEmpty()) {
+            stem = literal;
+            if (stem.endsWith("/") && stem.length() > 1) {
+                //we don't allow stem to end with a slash
+                //so if have /hello and /hello/ they can both be matched
+                //all JAX-RS paths have an implicit (/.*)? at the end of them
+                //to technically every path has an optional implicit slash
+                stem = stem.substring(0, stem.length() - 1);
+                components.add(new TemplateComponent(Type.LITERAL, stem, null, null, null));
+                components.add(new TemplateComponent(Type.LITERAL, "/", null, null, null));
+            } else {
+                components.add(new TemplateComponent(Type.LITERAL, literal, null, null, null));
+            }
+        } else {
+            components.add(new TemplateComponent(Type.LITERAL, literal, null, null, null));
+        }
+        return stem;
     }
 
     public URITemplate(String template, String stem, int literalCharacterCount,

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/GeneralResource.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/GeneralResource.java
@@ -1,0 +1,18 @@
+package org.jboss.resteasy.reactive.server.vertx.test.matching;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/hello")
+public class GeneralResource {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    @Path("{id}")
+    public String hello(@PathParam("id") String id) {
+        return "general:" + id;
+    }
+}

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/SpecificResource.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/SpecificResource.java
@@ -1,0 +1,16 @@
+package org.jboss.resteasy.reactive.server.vertx.test.matching;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/hello/{id}/bar")
+public class SpecificResource {
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello(@PathParam("id") String id) {
+        return "specific:" + id;
+    }
+}

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/StemMatchTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/StemMatchTest.java
@@ -1,0 +1,39 @@
+package org.jboss.resteasy.reactive.server.vertx.test.matching;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.CoreMatchers.is;
+
+import java.util.function.Supplier;
+import org.jboss.resteasy.reactive.server.vertx.test.framework.ResteasyReactiveUnitTest;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class StemMatchTest {
+
+    @RegisterExtension
+    static ResteasyReactiveUnitTest test = new ResteasyReactiveUnitTest()
+            .setArchiveProducer(new Supplier<JavaArchive>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(SpecificResource.class, GeneralResource.class);
+                }
+            });
+
+    @Test
+    public void testPathMatchSpecifics() {
+        given()
+                .when().get("/hello/foo")
+                .then()
+                .statusCode(200)
+                .body(is("general:foo"));
+        given()
+                .when().get("/hello/foo/bar")
+                .then()
+                .statusCode(200)
+                .body(is("specific:foo"));
+    }
+
+}

